src/http/common.cr

```diff
@@ -229,6 +229,85 @@ module HTTP
     time.to_utc.to_s("%a, %d %b %Y %H:%M:%S GMT")
   end
 
+  # Dequotes an [RFC 2616](https://tools.ietf.org/html/rfc2616#page-17)
+  # quoted-string.
+  #
+  # ```
+  # quoted = %q(\"foo\\bar\")
+  # HTTP.dequote_string(quoted) # => %q("foo\bar")
+  # ```
+  def self.dequote_string(str)
+    data = str.to_slice
+    quoted_pair_index = data.index('\\'.ord)
+    return str unless quoted_pair_index
+
+    String.build do |io|
+      while quoted_pair_index
+        io.write(data[0, quoted_pair_index])
+        io << data[quoted_pair_index + 1].chr
+
+        data += quoted_pair_index + 2
+        quoted_pair_index = data.index('\\'.ord)
+      end
+      io.write(data)
+    end
+  end
+
+  # :nodoc:
+  QUOTE_LOOKUP_TABLE = begin
+    # 0 = direct
+    # 1 = backslash escape
+    # 2 = raise
+    table = Pointer(UInt8).malloc(256, 0_u8)
+
+    # See https://evolvis.org/pipermail/evolvis-platfrm-discuss/2014-November/000675.html
+
+    (0x00..0x1F).each { |byte| table[byte] = 2_u8 } # All Control characters raise
+    table[0x09] = 1_u8                              # Except tab, which is escaped.
+
+    table[0x20] = 1_u8 # Space is escaped,
+    table[0x22] = 1_u8 # Quotes are escaped, and
+    table[0x5C] = 1_u8 # Backslash is escaped.
+    table[0x7F] = 2_u8 # DEL is a control character, so also raises.
+
+    table
+  end
+
+  # Encodes a string to an [RFC 2616](https://tools.ietf.org/html/rfc2616#page-17)
+  # quoted-string. Encoded string is written to *io*. May raise when *string*
+  # contains an invalid character.
+  #
+  # ```
+  # string = %q("foo\ bar")
+  # io = IO::Memory.new
+  # HTTP.quote_string(string, io)
+  # io.gets_to_end # => %q(\"foo\\\ bar\")
+  # ```
+  def self.quote_string(string, io)
+    string.each_byte do |byte|
+      case QUOTE_LOOKUP_TABLE[byte]
```

Comment:
We think it's better here to just do `case` on the byte, without a lookup table. The code will be clearer and it won't be unsafe (because it uses Pointer).